Add evaluate flag to rewrites that return unevaluated expressions

asmeurer opened the pull request:
Setting it to True (it is False by default) disables rewriting cases that create unevaluated Pow/exp objects. This also applies to a few other places in the codebase.

This is used in solve to avoid a recursion error (fixes #24368) which is caused by the call to rewrite(exp) in solve, which then gets rewritten back to a Pow by Poly. There doesn't appear to be any benefit to solving to rewrite powers with nonsymbolic exponents to exp (would love confirmation on this).

<!-- Your title above should be a short description of what
was changed. Do not include the issue number in the title. -->

#### References to other Issues or PRs
<!-- If this pull request fixes an issue, write "Fixes #NNNN" in that exact
format, e.g. "Fixes #1234" (see
https://tinyurl.com/auto-closing for more information). Also, please
write a comment on that issue linking back to this pull request once it is
open. -->


#### Brief description of what is fixed or changed


#### Other comments


#### Release Notes

<!-- Write the release notes for this release below between the BEGIN and END
statements. The basic format is a bulleted list with the name of the subpackage
and the release note for this PR. For example:

* solvers
  * Added a new solver for logarithmic equations.

* functions
  * Fixed a bug with log of integers.

or if no release note(s) should be included use:

NO ENTRY

See https://github.com/sympy/sympy/wiki/Writing-Release-Notes for more
information on how to write release notes. The bot will check your release
notes automatically to see if they are formatted correctly. -->

<!-- BEGIN RELEASE NOTES -->
- solvers
  - Fix a recursion error when solving some modular equations.
- core
  - Add `evaluate` flag to various rewrites that produce unevaluated expressions (such as `Pow.rewrite(exp)` for nonsymbolic exponent). It is False by default but setting it to True prevents `rewrite` from returning an unevaluated expression.
<!-- END RELEASE NOTES -->

asmeurer committed:
Add a flag nonsymbolic_exponent to Pow.rewrite(exp)

Setting it to False (it is True by default) disables rewriting cases that
create unevaluated Pow/exp objects.

This is used in solve to avoid a recursion error (fixes #24368) which is
caused by the call to rewrite(exp) in solve, which then gets rewritten back to
a Pow by Poly. There doesn't appear to be any benefit to solving to rewrite
powers with nonsymbolic exponents to exp.

asmeurer commented:
Maybe the flag should instead be something more general about returning an unevaluated expression. That could be relevant to multiple rewrite methods.

asmeurer commented:
Looks like there's already precedent for an `evaluate` flag in `Relational.rewrite(Add)`.

asmeurer committed:
Use the evaluate flag in rewrite whenever it returns an unevaluated expression

The default is False (except in Relational.rewrite(Add), where it already
existed and defaulted to True), but this allows setting it to True to prevent
rewrite from returning unevaluated expressions. This is useful for library
functions that want to do rewrites but want to avoid creating unevaluated
expressions.

asmeurer commented:
OK, forget about that other flag name, I'm calling it `evaluate`. I also added it to other `rewrite` functions that can return unevaluated expressions. It defaults to False, but setting it to True prevents rewrite from returning an unevaluated expression.

I also replaced the `for_rewrite` flag in `doit` in the probability module since it was just the same thing as `not evaluate`.

asmeurer committed:
Pass **kwargs through recursive rewrite calls in trigonometric.py

asmeurer committed:
Note that **hints should be passed through to recursive rewrite calls in the custom functions guide

asmeurer committed:
Add tests to rewrite(evaluate=True) in the trig functions

smichr commented:
> I also replaced the `for_rewrite` flag in `doit` in the probability module since it was just the same thing as `not evaluate`.

This is not backwards compatibile, isit? But are you assuming that almost nobody is using this?

asmeurer commented:
> This is not backwards compatibile, isit? But are you assuming that almost nobody is using this?

It wasn't documented anywhere, and it seems pretty clear to me from the flag name that it was only ever intended for internal use.

asmeurer commented:
By the way, my main worry here is whether disabling the rewriting of powers with non-symbolic exponent to exp in solve() could result in a regression. I don't actually know what sorts of exponential equations solve can handle.

smichr commented:
> I don't actually know what sorts of exponential equations solve can handle.

If tests are passing I wouldn't worry about it too much. At worst it will mean that some solver will have to pass a different hint to the rewrite.

smichr commented:
As I recall, that attempt of rewriting in the solvers is trying to rewrite trigonometric functions in terms of exponentials to try get a polynomial in exponentials, e.g. `(cos(x)+sin(x)-S.Half).rewrite(exp)` to get solutions.

smichr reviewed (approved):
Flexibilty of allowing rewrite to be evaluated or not seems good, especially in light of automatic rewriting. The use of rewrite(exp) in solve was intended to handle trigonometric functions, not turn things like `x**2` into `exp(2*log(x)`, so I don't think there is a conflict of intent there.

oscarbenjamin commented:
I expect that a lot of users will be confused by this. There are a few cases where rewrite uses `evaluate=False`. There are far more cases where a rewrite expression will evaluate regardless of any evaluate keyword argument.

The OP and release note say that the evaluate flag is "False by default" but this is probably not how most people would understand this. I think that most people who would be interested in having an `evaluate` flag for `rewrite` would want to use it to set `evaluate=False` in order to prevent evaluation e.g.:
```python
In [2]: cosh(x).rewrite(cos)
Out[2]: cosh(x)

In [3]: cosh(x).rewrite(cos, evaluate=False)
Out[3]: cosh(x)

In [4]: cos(I*x)
Out[4]: cosh(x)

In [5]: cos(I*x, evaluate=False)
Out[5]: cos(ⅈ⋅x)
```
Is the intention here that `evaluate=False` should work for all rewrites?

I imagine that more people would be interested in having `evaluate=False` work than `evaluate=True`. In fact if `evaluate=False` was an option we could change the rewrites that currently use `evaluate=False` to have behaviour respecting the evaluate flag (and have it default to True). I expect that the only reason these methods use `evaluate=False` now is to work around the limitation that it is not possible to meaningfully pass `evaluate=False` through `rewrite`.

oscarbenjamin commented:
> This is used in solve to avoid a recursion error (fixes #24368) which is caused by the call to rewrite(exp) in solve, which then gets rewritten back to a Pow by Poly.

For the issue in `solve` I think a different approach should be used. I've seen this issue come up in many different examples and it's really a bug in solve.

asmeurer commented:
> The OP and release note say that the evaluate flag is "False by default" but this is probably not how most people would understand this. I think that most people who would be interested in having an evaluate flag for rewrite would want to use it to set evaluate=False in order to prevent evaluation e.g.:

I would say that `cosh(x).rewrite(cos)` is currently bugged, and it ought to be using evaluate=False (I can add that here). The idea here is that anything in a rewrite that would evaluate away ought to be using evaluate=False by default, but this flag allows overriding that so that you can do rewrites in library code without getting broken unevaluated expressions. 

> I imagine that more people would be interested in having evaluate=False work than evaluate=True.

That's the point, that evaluate=False is the default for rewrites (where it matters). The nondefault here is evaluate=True.

Originally I had the default as evaluate=None, but I found I was just writing

```py
if evaluate is None:
    evaluate = False
```

which indicated to me that the default is really False (the actual default in the signature doesn't really matter because it's hidden in the specific `_eval_rewrite` signature; actually I didn't even document this anywhere publicly outside of the release note).

asmeurer commented:
> For the issue in solve I think a different approach should be used. I've seen this issue come up in many different examples and it's really a bug in solve.

I guess you mean that solve shouldn't be recursive at all? That seems like a much harder fix.

oscarbenjamin commented:
> I guess you mean that solve shouldn't be recursive at all?

I've commented in the issue: https://github.com/sympy/sympy/issues/24368#issuecomment-1397489964

oscarbenjamin commented:
> The idea here is that anything in a rewrite that would evaluate away ought to be using evaluate=False by default

If you take a look at the codebase (`git grep -B3 _eval_rewrite_as`) you'll see that there are a lot of rewrite methods and almost none of them does anything to handle evaluate=False. This PR seems to be adding an undocumented `evaluate` parameter to `Basic.rewrite`. We need to consider carefully that this works as intended.

asmeurer commented:
rewrite already handles arbitrary hint keyword arguments. Every `_eval_rewrite` method should accept a `**hints` parameter, and ignore unknown hints. I see there's a few instances that aren't doing this correctly but they should be fixed.

asmeurer commented:
In general, I see this as more of a workaround for the fact that none of these things like `exp(2*log(x))` or `sin(pi - x)` should be evaluating in the first place. The fact that there are rewrite methods that try to return them is a sign that you'd want to be able represent them directly. The only exception is in `Relational.rewrite(Add)`, where the default for `evaluate` already isn't False.

asmeurer commented:
If we can fix the solve bug in another way, I'm OK with removing the `evaluate` flag here, as it would otherwise not have a use-case. I think some of the other cleanups in this PR are still good, though.

asmeurer committed:
Always include **kwargs in _eval_rewrite_as_ methods

This allows arbitrary hints to be passed through from rewrite() (similar to
doit()).

asmeurer committed:
Remove debug code

asmeurer committed:
Use evaluate=False in hyperbolic trig function rewrites

oscarbenjamin commented:
I'm not opposing the change to `evaluate` but just pointing out that introducing this has other consequences. With this PR right now an undocumented parameter is introduced that sounds a lot like what a lot of people want except that you've introduced it for the opposite purpose so it doesn't actually do what they want.

> rewrite already handles arbitrary hint keyword arguments

I don't think that many users would perceive `evaluate` as just being a "hint" here. It's important to understand the difference in perspective between you who want to avoid "broken unevaluated expressions" and many users who actually want to work with unevaluated expressions and don't understand why SymPy doesn't just support them properly.

asmeurer committed:
Fix failing test

smichr commented:
> If we can fix the solve bug in another way,

Can't we just guard teh exp rewrite with a test that the expression has trigonometric (regular or hyper) in a symbol of interest?

asmeurer commented:
> Can't we just guard the exp rewrite with a test that the expression has trigonometric (regular or hyper) in a symbol of interest?

That's what I did at #24553, although Oscar has a suggestion for a more involved improvement at https://github.com/sympy/sympy/issues/24368#issuecomment-1397731889 (would love to hear your thoughts on how straightforward that would be to implement).

asmeurer commented:
To paraphrase from https://github.com/sympy/sympy/pull/24553, the targeted rewrite leads to various test failures. It seems that some solves actually do rely on rewriting powers to exp, not just trig functions. Trying to keep these solves working by other means has led down a rabbit hole that I personally do not know how to get out of. 

So I'm feeling that this PR is the better solution for now. It keeps the existing solve functionality and logic, as convoluted as it is, and just prevents the one case that is definitely wrong (rewriting powers that are so trivial that they just gets evaluated back).

asmeurer committed:
Merge branch 'master' into solve-exp-rewrite-fix

asmeurer commented:
Thoughts on this PR @smichr or @oscarbenjamin (see my previous comment)?

oscarbenjamin commented:
The idea here seems to be that `evaluate` can True, False or None (unspecified) and that some rewrite methods would differ between `evaluate=None` and `evaluate=True`.

A user would probably be more interested in being able to use `evaluate=False` like:
```python
In [22]: atan2(1, 1, evaluate=False)
Out[22]: atan2(1, 1)

In [23]: atan2(1, 1, evaluate=False).rewrite(tan)
Out[23]: 
π
─
4

In [24]: atan2(1, 1, evaluate=False).rewrite(tan, evaluate=False)
Out[24]: 
π
─
4

In [25]: atan(1, evaluate=False)
Out[25]: atan(1)

In [26]: atan(1)
Out[26]: 
π
─
4
```
Do we want to support `rewrite(evaluate=False)` and if so what should its expected behaviour be?

Is it expected that every rewrite method should handle `evaluate=False`?

Do we really want there to be three cases for `evaluate` and if so when should the different cases apply?

asmeurer commented:
I hadn't considered an example like that. I made it so that the evaluate flag to rewrite manages the instances where evaluate=False is the default, but I hadn't considered making it also apply to cases where evaluate=True is the default. 

Maybe a cleaner option here would be to manage the flag in the top-level `rewrite` and just use `with evaluate(False)` when it's set. Of course, the methods that always return unevaluated expressions should still do evaluate=False when evaluate=None.

asmeurer committed:
Make the evaluate context manager reusable and reentrant

asmeurer commented:
I guess a downside there is that evaluate(False) applies to every possible expression. I implemented this idea, and this test fails:

```
    eq = Eq(x + y, y - x)
    assert eq.rewrite(Add, evaluate=False).args == (x, y, x, -y)
```

because the `args` are now actually `(x, y, -(-x + y))` because it evaluate=False prevents the Add that is created from denesting. 

I guess we could say that sort of behavior is really just the expected behavior of setting evaluate=False in anything. 

By the way there's no `evaluate=None` here (except in Eq.rewrite_Add, which already has it in master). Some rewrite methods have `evaluate=False` as the default, which is used if `evaluate` is not passed at all. That means they default to creating *specific* expressions in an unevaluated way, namely, the ones that would just evaluate away the rewrite if they weren't (like `sin(pi - x)`).

asmeurer commented:
It occurs to me that if you're still uneasy about the broader changes to `rewrite` here, a simpler fix would be to just do

```diff
diff --git a/sympy/solvers/solvers.py b/sympy/solvers/solvers.py
index fcc7e13c31..c1bff685a0 100644
--- a/sympy/solvers/solvers.py
+++ b/sympy/solvers/solvers.py
@@ -42,6 +42,7 @@
     separatevars)
 from sympy.simplify.sqrtdenest import sqrt_depth
 from sympy.simplify.fu import TR1, TR2i
+from sympy.strategies.rl import rebuild
 from sympy.matrices.common import NonInvertibleMatrixError
 from sympy.matrices import Matrix, zeros
 from sympy.polys import roots, cancel, factor, Poly
@@ -2770,6 +2771,7 @@ def equal(expr1, expr2):
                 return _vsolve(lhs.args[0] - rhs*exp(rhs), sym, **flags)
 
         rewrite = lhs.rewrite(exp)
+        rewrite = rebuild(rewrite)
         if rewrite != lhs:
             return _vsolve(rewrite - rhs, sym, **flags)
     except NotImplementedError:
```

This has the same effect for the actual fix in solve().

For what it's worth though, the rewrite changes here actually fix many rewrites that aren't actually working, e.g., `sech(x).rewrite(sinh)`:

**master**

```py
>>> sech(x).rewrite(sinh)
1/cosh(x)
```

**This branch**

```py
>>> sech(x).rewrite(sinh)
I/sinh(x + I*pi/2)
```

(and there are dozens of similar examples, see the tests changes in the diff of this PR). 

Perhaps we can sidestep the concept of allowing the user to pass `evaluate` to rewrite, but still make sure the above rewrites always unconditionally use evaluate=False, and if they want to evaluate them, they can use `rebuild` or something similar.

oscarbenjamin commented:
> This has the same effect for the actual fix in solve().

I think that the solve problem is fixed just by restricting which types get rewritten isn't it?

asmeurer commented:
It isn't. See the issues I ran into in https://github.com/sympy/sympy/pull/24553. Some solvers really do use the Pow -> exp rewrite, and trying to work around that has led me down too many rabbit holes and I've given up trying to figure out how to make everything work. Instead of trying to whitelist the rewrites that are used, it seems to be much simpler to just avoid the one rewrite that we know is problematic and isn't actually used (i.e., the one that just evaluates back to Pow anyway).

oscarbenjamin commented:
> By the way there's no `evaluate=None` here

We have some rewrites that do different things dependent on whether `evaluate=True/False/None`. In some cases None is equivalent to True and in other it is equivalent to False. From the caller's perspective that means they should choose between three possible values that might affect the rewrite of different parts of the expression in different ways.

I don't necessarily object to this change but it needs to be thought through fully including considering all possible values of evaluate and specifying how all rewrite functions should be written in future to match what expectations users should have.

asmeurer committed:
Make rewrite(evaluate=False) apply evaluate=False to all subexpressions

asmeurer commented:
Well anyway, I think I'm convinced that maybe this isn't such a great idea. So what I think I'm going to do is:

- Open a new PR with just the patch from https://github.com/sympy/sympy/pull/24549#issuecomment-1492639098 to fix the original solve() issue. I just confirmed that all the tests pass with it. This is definitely the simplest possible fix for the issue.

- Open another PR or two for the independent rewrite fixes from this PR (specifically, getting things like `sech(x).rewrite(sinh)` to work), but without adding an `evaluate` flag to rewrite (the flag will still be present for `Eq.rewrite(Add)`, which is itself already [problematic for other reasons](https://github.com/sympy/sympy/issues/24650#issuecomment-1435471545)).

asmeurer commented:
See https://github.com/sympy/sympy/pull/24999

asmeurer commented:
Going to close this in favor of those other two PRs.
